src/client/datascience/context/telemetry.ts

```diff
@@ -0,0 +1,320 @@
+import { JupyterSelfCertsError } from '../jupyter/jupyterSelfCertsError';
+import { Telemetry } from '../constants';
+import { WorkspaceInterpreterTracker } from './workspaceInterpreterTracker';
+import { InterruptResult } from '../types';
+import { getResourceType, getTelemetrySafeLanguage } from '../common';
+import { PYTHON_LANGUAGE } from '../../common/constants';
+import { InterpreterCountTracker } from './interpreterCountTracker';
+import { FetchError } from 'node-fetch';
+
+type ContextualTelemetryProps = {
+    kernelConnection: KernelConnectionMetadata;
+    /**
+     * Used by WebViews & Interactive window.
+     * In those cases we know for a fact that the user changes the kernel.
+     * In Native Notebooks, we don't know whether the user changed the kernel or VS Code is just asking for default kernel.
+     * In Native Notebooks we track changes to selection by checking if previously selected kernel is the same as the new one.
+     */
+    kernelConnectionChanged: boolean;
+    startFailed: boolean;
+    kernelDied: boolean;
+    interruptKernel: boolean;
+    restartKernel: boolean;
+    kernelSpecCount: number; // Total number of kernel specs in list of kernels.
+    kernelInterpreterCount: number; // Total number of interpreters in list of kernels
+    kernelLiveCount: number; // Total number of live kernels in list of kernels.
+};
+
+type Context = {
+    previouslySelectedKernelConnectionId: string;
+};
+const trackedInfo = new Map<string, [ResourceSpecificTelemetryProperties, Context]>();
+const currentOSType = getOSType();
+
+export function getErrorClassification(error: Error) {
+    if (error.message.indexOf('reason: self signed certificate') >= 0) {
+        return 'jupyterselfcert';
+    } else if (isErrorType(error, JupyterSelfCertsError)) {
+        return 'jupyterselfcert';
+    } else if (isErrorType(error, JupyterWaitForIdleError)) {
+        return 'timeout';
+    } else if (isErrorType(error, TimedOutError)) {
+        return 'timeout';
+    } else if (isErrorType(error, JupyterInvalidKernelError)) {
+        return 'invalidkernel';
+    } else if (isErrorType(error, JupyterKernelPromiseFailedError)) {
+        return 'kernelpromisetimeout';
+    } else if (isErrorType(error, IpyKernelNotInstalledError)) {
+        return 'noipykernel';
+    } else if (isErrorType(error, CancellationError)) {
+        return 'cancelled';
+    } else if (isErrorType(error, JupyterSessionStartError)) {
+        return 'jupytersession';
+    } else if (isErrorType(error, JupyterConnectError)) {
+        return 'jupyterconnection';
+    } else if (isErrorType(error, JupyterInstallError)) {
+        return 'jupyterinstall';
+    } else if (isErrorType(error, KernelDiedError)) {
+        return 'kerneldied';
+    } else if (isErrorType(error, FetchError)) {
+        return 'fetcherror';
+    }
+    return 'unknown';
+}
+export function sendKernelTelemetryEvent<P extends IEventNamePropertyMapping, E extends keyof P>(
```

Comment:
I'm worried nobody will know when to sendKernelTelemetryEvent versus a sendTelemetryEvent

Could we send the different event based on the eventName?

Reply:
Meaning sendTelemetryEvent would know whether or not to send all the extra kernel data based on the event name instead of having the caller determine whether or not to call sendKernelTelemetryEvent or sendTelemetryEvent.

Reply:
We can, but then I'd have to modify the `sendTelemetryEvent` method to overload it with the `resource` arguments.
& you do raise a very good point.
How about we re-visit that in a subseparate PR.
I do agree it will not get used much & it needs to be fixed. But feels a bigger change, hence would prefer to defer it.